invoke-ai/InvokeAI
# [Feature] Enable Stable Fast acceleration

## What type of PR is this? (check all applicable)



## Have you discussed this change with the InvokeAI team?

      
## Have you updated all relevant documentation?
- [X] Not yet!


## Description

This is a proof-of-principle demonstration of integrating the [Stable-Fast](https://github.com/chengzeyi/stable-fast)  model compiler into InvokeAI. In conjunction with xformers, Stable-Fast roughly doubles performance on denoising and vae decoding. The main drawback is that there is a noticeable compilation delay the first time a model is used. Rendering pauses for 5 seconds or so while the model is compiled and "warms up", resulting in a few slow preview frames before the acceleration kicks in. After this, rendering is noticeably faster: I see about a 60% improvement on step rendering.

LoRAs, Controlnets T2Is, TI embeddings and IP Adapters are all working as expected. However, I had to do a little hacking to get IP Adapters to work: please see `invokeai/backend/ip_adapter/ip_attention_weights.py` for where I manually set `requires_grad` to False for all the `ip_attention_weights` parameters.

***Bug:*** You can’t change the weight of an IP Adapter prompt after the first generation. The weight gets stuck at whatever it was the first time the image prompt is used, presumably because the JIT compiler kicks in after the IP Adapter tensors are patched in. LoRAs are fine, however, so there is some difference in the model patching mechanism.


## Related Tickets & Documents



- Related Issue #
- Closes #

## QA Instructions, Screenshots, Recordings

Run `pip install -e .[xformers,stable-fast]` to install the stable-fast library. Then launch `invokeai-web`, select a model and render. Everything should "just work."



## Merge Plan

I will merge this when approved. I will tidy up the commit history beforehand.



## Added/updated tests?

- [X] Not yet

## [optional] Are there any post deployment tasks we need to perform?
